vcf-split.1
.TH VCF-SPLIT 1
.SH NAME    \" Section header
.PP
 
vcf-split \- Efficiently split a multi-sample VCF stream into single-sample files

\" Convention:
\" Underline anything that is typed verbatim - commands, etc.
.SH SYNOPSIS
.PP
.nf 
.na 
vcf-split \\
    [--het-only] [--alt-only] [--max-calls N] \\
    [--sample-id-file file] [--output-fields field-spec] \\
    output-file-prefix first-column last-column < file.vcf

bcftools view file.bcf | vcf-split ...
.ad
.fi

.SH OPTIONS and ARGUMENTS
.TP
\fB\-\-het\-only
Output only heterozygous sites.  When decoding a BCF file,
"bcftools view --genotype het" slows down the "bcftools view" process
and vcf-split uses far less CPU.  Since bcftools is already saturating a
CPU core and vcf-split has CPU cycles to spare, allowing vcf-split
to perform the heterozygous site selection increases pipeline performance
considerably.

.TP
\fB\-\-alt\-only
Output only sites with at least one ALT allele.

.TP
\fB\-\-max\-calls N
Limit the number of VCF calls processed (for quick testing without the need to
generate smaller test input files).

.TP
\fB\-\-sample\-id\-file filename
File containing a whitespace-separated list of arbitrary sample IDS, which
must match the sample names in the VCF input.

.TP
\fB\-\-output\-fields field-spec
Indicates which fields to pass to the output.  Fields not indicated here are
replaced with a reasonable placeholder for that field, such as ".".
field-spec is a comma-separated list of fields to include in the output
including one or more of chrom,pos,id,ref,alt,qual,filter, and info.

.TP
.B output-file-prefix
Common filename prefix for all single-sample output files (see Examples
directory).

.TP
.B first-column last-column
1-based column numbers limiting the number of samples for each run.
.B vcf-split opens one output stream for each sample and many systems
cannot support more than 30,000 to 40,000 open files at a time.
E.g. for a 100,000 sample VCF stream, you may want to do multiple runs of
10,000 each (see EXAMPLES below).  To prevent system overload, the maximum
number of open files is hard-coded at 10,000.  To override this limit, you
must edit vcf-split.h and recompile.  Note that using
--sample-id-file may limits the number of open files to less than
last-column - first_column + 1 and may make multiple runs unnecessary.

\" Optional sections
.SH "PURPOSE"

.B vcf-split
efficiently splits a multi-sample VCF stream into single-sample VCF files.

.SH "DESCRIPTION"

Traditional methods for splitting a multi-sample VCF stream into single-sample
files involve a loop or parallel job that rereads the multi-sample input for
every sample.  This is grossly inefficient and can
become a major bottleneck where there are many samples and/or the input
is compressed.  For example, using "bcftools view" with optimal filtering
options to decode one human chromosome BCF with
137,977 samples and pipe the VCF output through "wc" took 12 hours on a
fast server using 2 cores.  To split it into 137,977 single-sample VCFs
would therefore require about 137,977 * 12 * 2 = ~3 million core-hours.
This translates to 171 years on a single server or 125 days using 1000 cores
on an HPC cluster, assuming file I/O is not a bottleneck with 1000 processes
reading the same input file.  ( The input would need to be prestaged on
multiple local disks to avoid overloading the network file system. )

.B vcf-split
solves this problem by writing a large number of single-sample VCFs
simultaneously during a single read of the multi-sample input.  Modern Unix
systems support tens of thousands of simultaneously open files, providing
a simple way to achieve enormous speedup.

To avoid system overload, vcf-split has a hard-coded limit of 10,000 samples
at a time.  Hence, vcf-split may reduce the time required to split a large
VCF by a factor of 10,000 and can process 137,977 samples in 14 passes.

.B vcf-split
is written entirely in C and attempts to optimize CPU, memory,
and disk access.  It does not inhale large amounts of data into RAM, so memory
use is trivial and it runs mostly in cache RAM, making computational code
as fast as possible.

The example BCF file mentioned above can be split in a few days on a single
server using the maximum of 10,000 samples per run.

.SH "SEE ALSO"
ad2vcf, vcf2hap, haplohseq, biolibc

.SH "EXAMPLES"

Split a simple VCF file with 100 samples, filtering for specific sample IDs:

.nf
.na
vcf-split < input.vcf --het-only --sample-id-file samples.csv \\
    single-sample- 1 100
.ad
.fi

Split a large BCF file with 120,000 samples (too many for your open file
limit):

.nf
.na
bcftools view --min-ac 2 --exclude-types indels \\
    freeze.8.chr1.pass_only.phased.bcf \\
    | vcf-split --het-only chr01. 1 30000

bcftools view --min-ac 2 --exclude-types indels \\
    freeze.8.chr1.pass_only.phased.bcf \\
    | vcf-split --het-only chr01. 30001 60000

bcftools view --min-ac 2 --exclude-types indels \\
    freeze.8.chr1.pass_only.phased.bcf \\
    | vcf-split --het-only chr01. 60001 90000

bcftools view --min-ac 2 --exclude-types indels \\
    freeze.8.chr1.pass_only.phased.bcf \\
    | vcf-split --het-only chr01. 90001 120000
.ad
.fi

.SH BUGS
Please report bugs to the author and send patches in unified diff format.
(Run "man diff" for more information)

.SH AUTHOR
.nf
.na
Jason W. Bacon
Paul Auer Lab
UW -- Milwaukee Zilber School of Public Health


vcf-split.c
/***************************************************************************
 *  Description:
 *      Generate single-sample VCF files suitable for haplohseq from
 *      dbGaP multisample BCFs.
 *
 *      Input is a VCF stream, usually piped in from bcftools view,
 *      containing multiple samples per call.  vcf-split outputs a
 *      single-sample VCF file for each sample in the columms specified
 *      on the command-line.  Output filenames are based on the sample
 *      ID for each column.
 *
 *  Arguments:
 *      first-column
 *      last-column
 *
 *  Returns:
 *      See "man sysexits"
 *
 *  History: 
 *  Date        Name        Modification
 *  2019-12-06  Jason Wayne BaconBegin
 ***************************************************************************/

#include <stdio.h>
#include <sysexits.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>     // PATH_MAX
#include <stdint.h>     // SIZE_MAX
#include <ctype.h>
#include <stdbool.h>
#include <errno.h>
#include <fcntl.h>
#include <unistd.h>
#include <vcfio.h>
#include "vcf-split.h"

int     main(int argc,const char *argv[])

{
    char        *field_spec,
		*eos;
    const char  *outfile_prefix,
		*selected_samples_file = NULL;
    id_list_t   *selected_sample_ids = NULL;
    size_t      first_col,
		last_col,
		max_calls = SIZE_MAX,
		next_arg = 1;
    flag_t      flags = 0;
    vcf_field_mask_t    field_mask = VCF_FIELD_ALL;
    
    next_arg = 1;
    while ( (next_arg < argc ) && (argv[next_arg][0] == '-') )
    {
	if ( strcmp(argv[next_arg],"--max-calls") == 0 )
	{
	    max_calls = strtoul(argv[++next_arg], &eos, 10);
	    if ( *eos != '\0' )
	    {
		fprintf(stderr, "%s: %s: Max calls must be a positive integer.\n",
			argv[0], argv[next_arg]);
		exit(EX_DATAERR);
	    }
	    ++next_arg;
	}
	
	else if ( strcmp(argv[next_arg],"--sample-id-file") == 0 )
	{
	    selected_samples_file = argv[++next_arg];
	    ++next_arg;
	    selected_sample_ids =
		read_selected_sample_ids(argv, selected_samples_file);
	    fprintf(stderr, "%zu selected samples.\n", selected_sample_ids->count);
	}
	
	/*
	 *  Adding filter flags like --genotype het to bcftools can slow
	 *  it down significantly.  Choosing flags that maximize the speed
	 *  of bcftools and filtering for heterozygous sites here will make
	 *  the analysis run much faster.
	 */
	
	else if ( strcmp(argv[next_arg], "--het-only") == 0 )
	{
	    flags |= FLAG_HET;
	    ++next_arg;
	}

	else if ( strcmp(argv[next_arg], "--alt-only") == 0 )
	{
	    if ( flags & FLAG_HET )
	    {
		fprintf(stderr,
		    "%s: --het-only and --alt-only are mutually exclusive.\n",
		    argv[0]);
		exit(EX_USAGE);
	    }
	    flags |= FLAG_ALT;
	    ++next_arg;
	}

	else if ( strcmp(argv[next_arg], "--output-fields") == 0 )
	{
	    ++next_arg;
	    field_spec = (char *)argv[next_arg++];
	    if ( (field_mask = vcf_parse_field_spec(field_spec))
		    == VCF_FIELD_ERROR )
		usage(argv);
	}
	
	else
	    usage(argv);
    }
    
    if ( next_arg > argc - 3 )
	usage(argv);

    outfile_prefix = argv[next_arg++];
    
    first_col = strtoul(argv[next_arg], &eos, 10);
    if ( *eos != '\0' )
    {
	fprintf(stderr, "%s: %s: Column must be a positive integer.\n",
		argv[0], argv[next_arg]);
	usage(argv);
    }
    
    last_col = strtoul(argv[++next_arg], &eos, 10);
    if ( *eos != '\0' )
    {
	fprintf(stderr, "%s: %s: Column must be a positive integer.\n",
		argv[0], argv[next_arg]);
	usage(argv);
    }
    
    if ( (first_col < 1) || (last_col < 1) || (first_col > last_col) )
    {
	fprintf(stderr, "%s: Columns must be >= 1 and last_col >= first_col.\n", argv[0]);
	usage(argv);
    }
    
    return vcf_split(argv, stdin, outfile_prefix, first_col, last_col,
		     selected_sample_ids, max_calls, flags, field_mask);
}


/***************************************************************************
 *  Description:
 *      Split a multisample VCF stream into single-sample files.
 *
 *  History: 
 *  Date        Name        Modification
 *  2019-12-06  Jason Bacon Begin
 ***************************************************************************/

int     vcf_split(const char *argv[], FILE *vcf_infile,
		  const char *outfile_prefix,
		  size_t first_col, size_t last_col,
		  id_list_t *selected_sample_ids, size_t max_calls,
		  flag_t flags, vcf_field_mask_t field_mask)

{
    char    inbuf[BUFF_SIZE + 1],
	    *all_sample_ids[last_col - first_col + 1];
    bool    selected[last_col - first_col + 1];
    size_t  c;
    FILE    *header_stream;
    
    // Input is likely to come from "bcftools view" stdout.
    // What is optimal buffering for a Unix pipe?  Benchmark several values.
    setvbuf(vcf_infile, inbuf, _IOFBF, BUFF_SIZE);
    header_stream = vcf_skip_header(vcf_infile);
    vcf_get_sample_ids(vcf_infile, all_sample_ids, first_col, last_col);

    /*
    fputs("All sample IDs:", stderr);
    for (c = 0; c < last_col - first_col + 1; ++c)
	fprintf(stderr, " %s", all_sample_ids[c]);
    putc('\n', stderr);
    */
    
    tag_selected_columns(all_sample_ids, selected_sample_ids, selected,
			 first_col, last_col);

    if ( selected_sample_ids != NULL )
    {
	fputs("Tagged sample IDs:", stderr);
	for (c = 0; c < selected_sample_ids->count; ++c)
	    fprintf(stderr, " %s", selected_sample_ids->ids[c]);
	fputc('\n', stderr);
    }
    
    write_output_files(argv, vcf_infile, header_stream, 
		       (const char **)all_sample_ids,
		       selected, outfile_prefix,
		       first_col, last_col, max_calls, flags, field_mask);
    
    return EX_OK;
}


/***************************************************************************
 *  Description:
 *      Write individual VCF files for each sample.
 *
 *  Arguments:
 *
 *  History: 
 *  Date        Name        Modification
 *  2019-12-06  Jason Bacon Begin
 ***************************************************************************/

void    write_output_files(const char *argv[], FILE *vcf_infile, FILE *header,
			    const char *all_sample_ids[], bool selected[],
			    const char *outfile_prefix,
			    size_t first_col, size_t last_col,
			    size_t max_calls, flag_t flags,
			    vcf_field_mask_t field_mask)

{
    size_t  columns = last_col - first_col + 1,
	    c;
    int     fd;
    FILE    *vcf_outfiles[columns];
    char    filename[PATH_MAX + 1],
	    file_format[129];
    
    // Open all output streams
    for (c = 0; c < columns; ++c)
    {
	if ( selected[c] )
	{
	    snprintf(filename, PATH_MAX, "%s%s.vcf",
		     outfile_prefix, all_sample_ids[c]);
	    /* Find a way to do this without thousands of xz processes
	    if ( flags & FLAG_XZ )
	    {
		snprintf(cmd, CMD_MAX, "xz --stdout > %s.xz",filename);
		vcf_outfiles[c] = popen(cmd, "w");
	    }
	    else
	    */
	    vcf_outfiles[c] = fopen(filename, "w");
	    if ( vcf_outfiles[c] == NULL )
	    {
		fprintf(stderr, "%s: Cannot create %s: %s.\n",
			argv[0], filename, strerror(errno));
		exit(EX_CANTCREAT);
	    }
	    
	    /*
	     *  Add basic header
	     *  FIXME: Add option to copy all/part of source header
	     */
	    
	    rewind(header);
	    fgets(file_format, 128, header);
	    if ( memcmp(file_format, "##fileformat", 12) == 0 )
		fputs(file_format, vcf_outfiles[c]);
	    fputs("#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSAMPLE\n",
		  vcf_outfiles[c]);
	}
    }

    // Heart of the program, split each VCF line across multiple files
    for (c = 0; split_line(argv, vcf_infile, vcf_outfiles, all_sample_ids,
			   selected, first_col, last_col, max_calls, flags,
			   field_mask);
			   ++c)
	;
    
    // Close all output streams
    for (c = 0; c < columns; ++c)
    {
	if ( selected[c] )
	{
	    /*
	    if ( flags & FLAG_XZ )
		pclose(vcf_outfiles[c]);
	    else
	    */
	    fclose(vcf_outfiles[c]);
	    /*
	     *  Touch a .done file to indicate completion.  Another script
	     *  can use this to determine which .vcf files are ready for
	     *  compression.
	     */
	    snprintf(filename, PATH_MAX, "%s%s.vcf.done",
		     outfile_prefix, all_sample_ids[c]);
	    if ( (fd = open(filename, O_CREAT | O_TRUNC | O_WRONLY, 0644)) != -1 )
		close(fd);
	    else
		fprintf(stderr, "%s: Warning: Could not create %s: %s.\n",
			argv[0], filename, strerror(errno));
	}
    }
    fprintf(stderr, "%s completed successfully.\n", argv[0]);
}


/***************************************************************************
 *  Description:
 *      Split a single input line from the multisample VCF.
 *
 *  History: 
 *  Date        Name        Modification
 *  2019-12-06  Jason Bacon Begin
 ***************************************************************************/

int     split_line(const char *argv[], FILE *vcf_infile, FILE *vcf_outfiles[],
		   const char *all_sample_ids[], bool selected[],
		   size_t first_col, size_t last_col, size_t max_calls,
		   flag_t flags, vcf_field_mask_t field_mask)

{
    static size_t   line_count = 0,
		    max_info_len = 0;
    size_t          c,
		    field_len;
    int             delimiter;
    static vcf_call_t      vcf_call = VCF_CALL_INIT;
    char            *genotype;
    
    /*
     *  Read in VCF fields
     */
    
    /* Declared as static: Allocate only once and reuse */
    if ( vcf_call.single_sample == NULL )
	vcf_call_init(&vcf_call, VCF_INFO_MAX_CHARS, VCF_FORMAT_MAX_CHARS,
		  VCF_SAMPLE_MAX_CHARS);
    genotype = vcf_call.single_sample;
    
    // Check max_calls here rather than outside in order to print the
    // end-of-run report below
    if ( (line_count < max_calls) && 
	 (vcf_read_static_fields(vcf_infile, &vcf_call) == VCF_OK) )
    {
	if ( (++line_count % 100 == 0) && isatty(fileno(stderr)) )
	    fprintf(stderr, "%zu\r", line_count);
	
	if ( VCF_INFO_LEN(&vcf_call) > max_info_len )
	    max_info_len = VCF_INFO_LEN(&vcf_call);
	
	// Skip columns before first_col
	for (c = 1; c < first_col; ++c)
	{
	    delimiter = tsv_skip_field(vcf_infile);
	    if ( delimiter == EOF )
	    {
		dump_line(argv, "split_line(): Hit EOF skipping fields before first_col.\n",
			  &vcf_call, line_count, c, first_col, all_sample_ids,
			  "Genotype not relevant");
		exit(EX_DATAERR);
	    }
	    else if ( delimiter == '\n' )
	    {
		fprintf(stderr, "%s: split_line(): Reached EOL before first_col.\n", argv[0]);
		fprintf(stderr, "Does your input really have %zu samples?\n",
			first_col);
		exit(EX_USAGE);
	    }
	}
	
	for (; (c <= last_col) && 
	       (delimiter = tsv_read_field(vcf_infile, genotype,
				VCF_SAMPLE_MAX_CHARS, &field_len)) != '\n';
				++c)
	{
	    if ( delimiter == EOF )
	    {
		dump_line(argv, "split_line(): Encountered EOF while reading genotype fields.\n",
			  &vcf_call, line_count, c, first_col, all_sample_ids,
			  genotype);
		exit(EX_DATAERR);
	    }
	    if ( selected[c - first_col] )
	    {
		if ( (flags == FLAG_NONE) ||
		     ((flags == FLAG_HET) && (genotype[0] != genotype[2])) ||
		     ((flags == FLAG_ALT) &&
		      ((genotype[0] == '1') || (genotype[2] == '1'))) )
		{
		    vcf_write_ss_call(vcf_outfiles[c - first_col],
					    &vcf_call, field_mask);
		    /*
		    fprintf(vcf_outfiles[c - first_col],
			    "%s\t%s\t%s\t%s\t%s\t.\t.\t.\t%s\t%s\n",
			    VCF_CHROMOSOME(&vcf_call), VCF_POS_STR(&vcf_call),
			    VCF_ID(&vcf_call), VCF_REF(&vcf_call),
			    VCF_ALT(&vcf_call), VCF_FORMAT(&vcf_call),
			    genotype);
		    */
		}
	    }
	}
	if ( c < last_col )
	{
	    fprintf(stderr, "%s: split_line(): Reached EOL before last_col.\n", argv[0]);
	    fprintf(stderr, "Does your input really have %zu samples?\n", last_col);
	    exit(EX_USAGE);
	}

	// If this wasn't the last sample in the line, skip to EOL
	if ( delimiter != '\n' )
	{
	    if ( tsv_skip_rest_of_line(vcf_infile) == EOF )
	    {
		dump_line(argv, "split_line(): Encountered EOF skipping fields after last_col.\n",
			  &vcf_call, line_count, c, first_col, all_sample_ids,
			  "Genotype not relevant");
		exit(EX_DATAERR);
	    }
	}
	return 1;
    }
    else
    {
	fprintf(stderr, "%s: split_line(): No more VCF calls.\n", argv[0]);
	fprintf(stderr, "Processed %zu multi-sample VCF calls.\n", line_count);
	fprintf(stderr, "Max info_len = %zu.\n", max_info_len);
	return 0;
    }
}


void    dump_line(const char *argv[], const char *message, 
		  vcf_call_t *vcf_call, size_t line_count, size_t col,
		  size_t first_col, const char *all_sample_ids[], char *genotype)

{
    fprintf(stderr, "%s: %s\n", argv[0], message);
    fprintf(stderr, "Input line: %zu\n", line_count);
    fprintf(stderr, "Column: %zu Sample ID: %s:\n",
	    col, all_sample_ids[col - first_col]);
    fprintf(stderr, "SS VCF: %s\t%s\t.\t%s\t%s\t.\t.\t.\t%s\t%s\n",
	    VCF_CHROMOSOME(vcf_call), VCF_POS_STR(vcf_call),
	    VCF_REF(vcf_call), VCF_ALT(vcf_call),
	    VCF_FORMAT(vcf_call), genotype);
}


/***************************************************************************
 *  Description:
 *      Read list of sample IDs from a file.
 *
 *  History: 
 *  Date        Name        Modification
 *  2019-12-20  Jason Bacon Begin
 ***************************************************************************/

id_list_t   *read_selected_sample_ids(const char *argv[],
				      const char *samples_file)

{
    id_list_t   *list;
    size_t      c;
    char        temp_id[SAMPLE_ID_MAX + 1];
    FILE        *fp;
    
    if ( (fp = fopen(samples_file, "r")) == NULL )
    {
	fprintf(stderr, "%s: Cannot open %s: %s\n", argv[0], samples_file,
		strerror(errno));
	exit(EX_NOINPUT);
    }
    
    if ( (list = (id_list_t *)malloc(sizeof(id_list_t))) == NULL )
    {
	fprintf(stderr, "%s: Cannot allocate sample_ids structure.\n", argv[0]);
	exit(EX_UNAVAILABLE);
    }
    
    /*
     *  With such a small file, the cleanest approach is read to count the
     *  IDS, do one malloc(), rewind and read again.
     */

    for (list->count = 0; read_string(fp, temp_id, SAMPLE_ID_MAX) > 0; ++list->count )
	;
    
    if ( (list->ids = (char **)malloc(list->count * sizeof(char **))) == NULL )
    {
	fprintf(stderr, "%s: Cannot allocate sample_ids list.\n", argv[0]);
	exit(EX_UNAVAILABLE);
    }
    
    rewind(fp);
    for (c = 0; c < list->count; ++c)
    {
	read_string(fp, temp_id, SAMPLE_ID_MAX);
	if ( (list->ids[c] = strdup(temp_id)) == NULL )
	{
	    fprintf(stderr, "%s: Cannot allocate sample id %zu.\n", argv[0], c);
	    exit(EX_UNAVAILABLE);
	}
    }
    fclose(fp);
    
    // List may already be sorted, so stay away from qsort, which has
    // worst-case performance on presorted data
    // List should not be huge, so minor issue here
    // heapsort() is not portable, however
    qsort(list->ids, list->count, sizeof(char *),
	(int (*)(const void *, const void *))strptrcmp);
    return list;
}


/***************************************************************************
 *  Description:
 *      Read a whitespace-separated string.
 *
 *  Returns:
 *      String length.
 *
 *  History: 
 *  Date        Name        Modification
 *  2019-12-23  Jason Bacon Begin
 ***************************************************************************/

size_t  read_string(FILE *fp, char *buff, size_t maxlen)

{
    size_t  c = 0;
    int     ch;
    
    while ( isspace(ch = getc(fp)) )
	;
    
    while ( (c < maxlen) && ! isspace(ch) && (ch != EOF) )
    {
	buff[c++] = ch;
	ch = getc(fp);
    }
    buff[c] = '\0';
    return c;
}


void    usage(const char *argv[])

{
    fprintf(stderr, "\nUsage: %s\n\t[--het-only]\n\t[--alt-only]\n\t"
		    "[--max-calls N]\n\t[--sample-id-file file]\n\t"
		    "[--output-fields field-spec]\n\toutput-file-prefix\n\t"
		    "first-column\n\tlast-column\n\n", argv[0]);
    fputs("Press return to continue...", stderr);
    getchar();
    fprintf(stderr, "\n--het-only indicates that only heterozygous fields are output.\n"
		    "Allowing vcf-split to perform this filtering is faster than doing\n"
		    "it in bcftools in some cases.\n\n"
		    "--alt-only indicates that only fields with at least one alt allele are output.\n\n"
		    "--max-calls limits the number of calls processed (for testing purposes).\n\n"
		    "--sample-id-file indicates a list of samples to extract.  Names must\n"
		    "match the column header in the input VCF.\n\n"
		    "field-spec is a comma-separated list of fields to include in the output\n"
		    "including one or more of\n\n"
		    "chrom,pos,id,ref,alt,qual,filter,info\n\n"
		    "fields not indicated are replaced with a '.'.\n\n"
		    "output-file-prefix is prepended to .vcf\n"
		    "It may include one or more subdirectories\n\n"
		    "first-column and last column indicate the range of samples to process\n"
		    "Output is the intersection of this range and --sample-id-file.\n\n"
		    );
    exit(EX_USAGE);
}


void    tag_selected_columns(char *all_sample_ids[],
			     id_list_t *selected_sample_ids, bool selected[],
			     size_t first_col, size_t last_col)

{
    size_t  c,
	    total_selected;
    
    // No --sample-id-file, select all samples
    if ( selected_sample_ids == NULL )
    {
	for (c = first_col; c <= last_col; ++c)
	    selected[c - first_col] = true;
	total_selected = last_col - first_col + 1;
    }
    else
    {
	for (c = first_col, total_selected = 0; c <= last_col; ++c)
	{
	    selected[c - first_col] =
		(bsearch(all_sample_ids + c - first_col, selected_sample_ids->ids,
		    selected_sample_ids->count, sizeof(char *),
		    (int (*)(const void *, const void *))strptrcmp)
		    != NULL);
	    if ( selected[c - first_col] )
		++total_selected;
	}
    }

    if ( total_selected > MAX_OUTFILES )
    {
	fprintf(stderr,
		"vcf-split:  Maximum output files = %u.  You have selected %zu.\n"
		"Increase MAX_OUTFILES and recompile if a %ux speedup is really not enough\n"
		"and you want to bludgeon your system.  Otherwise, you can process the\n"
		"multi-sample input file in %u-sample chunks.\n",
		MAX_OUTFILES, total_selected, MAX_OUTFILES, MAX_OUTFILES);
	exit(EX_USAGE);
    }
}
